### PR TITLE
Add version digest as lambda env var

### PR DESCRIPTION
**Pull request by Brent-Jackson**
This change embeds an environment variable onto each lambda function with the value of the version config hash, ensuring that anytime the config hash changes, the resulting lambda configuration will be changed, meaning the Lambda::Version should always be successfully published.

Closes: https://github.com/serverless/serverless/issues/8251

### CODE REVIEW

**Comment by Brent-Jackson** [comment]
Current changes break most of the `AwsCompileFunctions` tests, but I wanted to open the PR to start implementation discussions before adjusting all the tests.

File: lib/plugins/aws/package/compile/functions/index.js [hunk 1]
```diff
@@ -474,6 +474,17 @@ class AwsCompileFunctions {
...
+            // Add an environment var with a value of the versionDigest to force
+            // the AWS::Lambda::Function configuration to change whenever the
+            // versionDigest changes
+            if (!functionResource.Properties.Environment) {
+              functionResource.Properties.Environment = {};
+            }
+            if (!functionResource.Properties.Environment.Variables) {
+              functionResource.Properties.Environment.Variables = {};
+            }
+            functionResource.Properties.Environment.Variables.SLS_LAMBDA_VERSION_DIGEST = versionDigest;
```
Comment (medikoo) [inline lib/plugins/aws/package/compile/functions/index.js:486]: Let's simplify this notation to:

```javascript
_.set(functionResource.Properties, 'Environment.Variables.SLS_LAMBDA_VERSION_DIGEST', versionDigest)
```

File: lib/plugins/aws/package/compile/functions/index.test.js [hunk 2]
```diff
@@ -2216,6 +2216,29 @@ describe('AwsCompileFunctions', () => {
         ).to.not.exist;
       });
     });
+
+    it('should only add an SLS_LAMBDA_VERSION_DIGEST env var to versioned functions', () => {
```
Comment (medikoo) [inline lib/plugins/aws/package/compile/functions/index.test.js:2220]: Ideally all new tests should be based on`runServerless` util. See https://github.com/serverless/serverless/tree/master/test#unit-tests

And in this file at the bottom we already have some tests that rely on that, let's add over there tests that confirms on that env var.

**Review by medikoo — CHANGES_REQUESTED** [review]
Thank you @Brent-Jackson that looks very good, I've proposed just few style improvements.

Therefore let's clean the tests, and have that in! :)

**Comment by Brent-Jackson** [comment]
Ah, looks like there are some integration tests I missed. I will tackle them tomorrow.

**Comment by Brent-Jackson** [comment]
@medikoo Tests are all fixed up. Not the cleanest, but I figured it's better to have less coupling for test cases rather than less code repetition.

**Review by medikoo — APPROVED** [review]
Thank you @Brent-Jackson !

Concerning testing, all new tests we introduce stand on `runServerless` util (see https://github.com/serverless/serverless/tree/master/test#unit-tests), this makes them less dependent on internals, and not require updates as ones we needed to patch here.

**Comment by evanuk** [comment]
This change breaks functions being associated to a CloudFront distribution, which requires the function to not have environment variables.
https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/lambda-requirements-limits.html#lambda-requirements-lambda-function-configuration

**Comment by Brent-Jackson** [comment]
> This change breaks functions being associated to a CloudFront distribution, which requires the function to not have environment variables.
> https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/lambda-requirements-limits.html#lambda-requirements-lambda-function-configuration

Hmm, that's unfortunate. 

I'm not particularly familiar with Lambda@Edge. Does the serverless framework currently have any notion of which lambdas will be "at edge"? i.e., could we just add an additional check to my logic to only embed the env var if the function is not "at edge"? Looking at https://www.serverless.com/blog/lambda-at-edge-support-added, it appears at a glance that we could just detect the presence of a `cloudFront` event type, and skip setting the env var. But this wouldn't work in a multi-stack situation where the Lambda and CloudFront distribution are defined in separate serverless applications. Not sure where to draw the line on whether or not that is in scope for the serverless framework.

Some alternatives that come to mind:
- Only add this env var if some additional configuration option is provided (I don't love this option since adding this env var is intended to be a bugfix, not a feature)
- Enforce any functions that will be run "at edge" to specify an additional configuration option to say so
  - This would be a breaking change
  - e.g.,
```yaml
functions:
  cfLambda:
    handler: functions/handler.cloudfront
    deployedAtEdge: true
```

**Comment by Brent-Jackson** [comment]
Actually, looking at the logic for compiling a `cloudFront` event, it seems the environment variables are removed entirely: 
https://github.com/serverless/serverless/blob/a020a4a683f7c5ef3625fc52cb319300b9e302d2/lib/plugins/aws/package/compile/events/cloudFront/index.js#L304-L317

That said, it still doesn't address the situation where the CloudFront integration is created with raw CloudFormation resources, and/or across stacks.

**Comment by evanuk** [comment]
Yeah, I imagine this use case is fairly common.  I have a lambda function in a stack that is used as CloudFront middleware in multiple other stacks.

**Comment by medikoo** [comment]
@evanuk great thanks for raising that. As @Brent-Jackson pointed we will not configure environment variables if lamba is expliclty attached to CloudFront distribution, but as you mention that's not good enough.

In light of that I will revert this PR, and let's move discussion on how to move forward to original issue #8251

**Pull request merged**